apps/redis.go

```diff
@@ -0,0 +1,102 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apps
+
+import (
+	"fmt"
+
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator"
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator/fluentbit"
+)
+
+type LoggingProcessorRedis struct {
+	confgenerator.ConfigComponent `yaml:",inline"`
+}
+
+func (LoggingProcessorRedis) Type() string {
+	return "redis"
+}
+
+func (p LoggingProcessorRedis) Components(tag string, uid string) []fluentbit.Component {
+	c := confgenerator.LoggingProcessorParseRegex{
+		// Documentation: https://github.com/redis/redis/blob/6.2/src/server.c#L1122
+		// Sample line: 534:M 28 Apr 2020 11:30:29.988 * DB loaded from disk: 0.002 seconds
+		// Sample older format line: [4018] 14 Nov 07:01:22.119 * Background saving terminated with success
+		Regex: `^\[?(?<pid>\d+):?(?<roleChar>[A-Z])?\]?\s+(?<time>\d{2}\s+\w+(?:\s+\d{4})?\s+\d{2}:\d{2}:\d{2}.\d{3})\s+(?<level>(\*|#|-|\.))\s+(?<message>.*)$`,
+		LoggingProcessorParseShared: confgenerator.LoggingProcessorParseShared{
+			TimeKey:    "time",
+			TimeFormat: "%d %b %Y %H:%M:%S.%L",
+			Types: map[string]string{
+				"pid": "integer",
+			},
+		},
+	}.Components(tag, uid)
+
+	for _, l := range []struct{ level, severity string }{
+		// Log levels documented: https://github.com/redis/redis/blob/6.2/src/server.c#L1124
+		{".", "DEBUG"},
+		{"-", "INFO"},
+		{"*", "NOTICE"},
+		{"#", "WARNING"},
+	} {
+		c = append(c, fluentbit.Component{
+			Kind: "FILTER",
+			Config: map[string]string{
+				"Name":      "modify",
+				"Match":     tag,
+				"Condition": fmt.Sprintf("Key_Value_Equals level %s", l.level),
+				"Add":       fmt.Sprintf("logging.googleapis.com/severity %s", l.severity),
+			},
+		})
+	}
+
+	for _, l := range []struct{ roleChar, role string }{
+		// Role translation documented: https://github.com/redis/redis/blob/6.2/src/server.c#L1149
+		{"X", "sentinel"},
+		{"C", "RDB/AOF_writing_child"},
+		{"S", "slave"},
+		{"M", "master"},
+	} {
+		c = append(c, fluentbit.Component{
+			Kind: "FILTER",
+			Config: map[string]string{
+				"Name":      "modify",
+				"Match":     tag,
+				"Condition": fmt.Sprintf("Key_Value_Equals roleChar %s", l.roleChar),
+				"Add":       fmt.Sprintf("role %s", l.role),
+			},
+		})
+	}
+	return c
+}
+
+type LoggingReceiverRedis struct {
+	LoggingProcessorRedis                   `yaml:",inline"`
+	confgenerator.LoggingReceiverFilesMixin `yaml:",inline" validate:"structonly"`
+}
+
+func (r LoggingReceiverRedis) Components(tag string) []fluentbit.Component {
+	if len(r.IncludePaths) == 0 {
+		r.IncludePaths = []string{"/var/log/redis/redis-server.log*", "/var/log/redis_*", "/var/log/redis/redis_*.log*"}
```

Comment:
Why does this have `redis-server.log*`? I imagine this will catch rotated log files, and possibly cause them to get processed a second time when they are renamed?

Reply:
Yeah, that seems like an oversight, I'll fix it.